pages/docs/guides/message-validation.md

```diff
@@ -0,0 +1,111 @@
+An example implementation of message validation in a gateway is the [AsyncAPI gateway](https://github.com/asyncapi/event-gateway). It intercepts all incoming messages moving them into a pipeline of middlewares and handlers such as message validation. You can use a Kafka consumer/producer[(kcat)](https://github.com/edenhill/kcat), a broker, and a simple WebSocket to run the AsyncAPI gateway in your machine.
+Check out an [AsyncAPI file demo with Studio](https://studio.asyncapi.com/?url=https://raw.githubusercontent.com/asyncapi/event-gateway/master/deployments/k8s/event-gateway-demo/event-gateway-demo.asyncapi.yaml) to learn more about how an AsyncAPI file can be used in gateway validation.
+
+<Remember>
+Currently, only the Kafka protocol is supported.
+</Remember>
+
+```mermaid
+graph TD
+    PR[Producer]-- Message --- EG[AsyncAPI Event-Gateway]
+    EG -- Message --- EGV{Is Message valid?}
+    EGV -->|Yes| BR[Broker]
+    EGV --- |No| INV{Fail when invalid?}
+    INV -->|Yes| ERR[/Fail/] -- Produce request errored --> PR
+    INV -->|No| BR
+```
+The AsyncAPI document is important because payload schemas are taken from it, and messages are validated against it in your application.
+You can spin up the AsyncAPI gateway using an AsyncAPI file. All the messages are forwarded to a WebSocket endpoint; if the message/payload is invalid, it includes a validation error message.
+
+### Schema Registry validation
+Producers and consumers do not communicate with each other directly; rather, information transfer happens via Kafka. At the same time, the consumer still needs to know the type of data the producer is sending. Imagine if the producer starts sending bad data to Kafka or if the data type of your data gets changed. We need a way to have a common data type that must be agreed upon.
+
+This is where Schema Registry comes into play. It is an application that runs outside your Kafka protocol and handles schema distribution to producers and consumers by storing a copy of the schema in its local cache and validating them in Kafka.
+
+```mermaid
+sequenceDiagram
+Producer ->> Schema Registry: Validate message schema
+Schema Registry ->> Producer: Return validation result
+
+Producer ->> Kafka: Publish message
+Kafka ->> Consumer: Recieves message
+Consumer ->> Schema Registry: Validate message schema
+Schema Registry ->> Consumer: Return validation result
+```
+
+With the Schema Registry in place, the producer first talks to the Schema Registry and checks if the schema is available before sending the data to Kafka. If it cannot locate the schema, it registers and caches it in the Schema Registry. When the producer receives the schema, it sends it to Kafka prefixed with a unique schema ID. When the consumer processes this message, it will communicate with the Schema Registry using the schema ID obtained from the producer. If there is a schema mismatch, the Schema Registry will throw an error, informing the producer that it violates the schema agreement.
```

Comment:
a small suggestion to make it more technically accurate

```suggestion
With the Schema Registry in place, the producer first talks to the Schema Registry and checks if the schema of the message it wants to send is available before sending it to the broker. If it cannot locate the schema, it registers it in the Schema Registry. Then the producer sends a message to the broker prefixed with a unique schema ID. When the consumer processes this message, it will communicate with the Schema Registry using the schema ID obtained from the producer. If there is a schema mismatch, the Schema Registry will throw an error, informing the producer that it violates the schema agreement.
```

you can remove below state diagram 👇🏼  as the sequence diagram you have above is excellent! and the state diagram is anyway not valid.